src/tools/upgrade-factory-snapshot.c

```diff
@@ -0,0 +1,725 @@
+static gchar *cmd_delete_snapshot[] = { SAILFISH_SNAPSHOT, "delete", SNAPSHOT_NAME, NULL };
+static gchar *cmd_deploy[] = { SAILFISH_SNAPSHOT, "deploy", SFMF_DEPLOY, SNAPSHOT_NAME, NULL };
+static gchar *cmd_delete_factory_rename[] = { SAILFISH_SNAPSHOT, "delete", FACTORY_RENAME, NULL };
+static gchar *cmd_rename_factory[] = { SAILFISH_SNAPSHOT, "rename", FACTORY_NAME, FACTORY_RENAME, NULL };
+static gchar *cmd_rename_snapshot[] = { SAILFISH_SNAPSHOT, "rename", SNAPSHOT_NAME, FACTORY_NAME, NULL };
+static gchar *sbj_partitions[] = { "@", "@home", NULL };
+
+static struct DeployTask g_deploy_tasks[] = {
+    { "Checking for existing snapshots", cmd_list, 1 },
+    { "Removing temporary snapshot", cmd_delete_snapshot, 0 },
+    { "Deploying new factory snapshot", cmd_deploy, 1 },
+    { "Removing renamed factory snapshot", cmd_delete_factory_rename, 0 },
+    { "Renaming factory snapshot", cmd_rename_factory, 1 },
+    { "Activating new factory snapshot", cmd_rename_snapshot, 1 },
+    { "Removing renamed factory snapshot", cmd_delete_factory_rename, 0 },
+    { "Enumerating snapshots", cmd_list, 0 },
+    { NULL, NULL, 0 },
+};
+
+static struct DeployTaskQueue g_deploy_queue = {
+    "deploy",
+
+    g_deploy_tasks,
+    -1,
+    -1,
+    0,
+
+    upgrade_factory_snapshot_finished_cb,
+    upgrade_factory_snapshot_task_done_cb,
+    NULL,
+};
+
+static struct DeployTask g_cleanup_tasks[] = {
+    { "Deleting temporary snapshot", cmd_delete_snapshot, 0 },
+    { "Enumerating snapshots", cmd_list, 0 },
+    { NULL, NULL, 0 },
+};
+
+static struct DeployTaskQueue g_cleanup_queue = {
+    "cleanup",
+
+    g_cleanup_tasks,
+    -1,
+    -1,
+    0,
+
+    NULL,
+    NULL,
+    NULL,
+};
+
+static struct UpgradeFactorySnapshot g_ufs = {
+    NULL,
+    NULL,
+    0,
+    0,
+    0,
+
+    FALSE,
+
+    &g_deploy_queue,
+    &g_cleanup_queue,
+
+    sbj_partitions,
+    -1,
+
+    { NULL, 0, NULL, 0, 0 },
+};
+
+static const GDBusInterfaceVTable upgrade_factory_snapshot_vtable = {
+    upgrade_factory_snapshot_method_call_cb,
+    upgrade_factory_snapshot_get_property_cb,
+    upgrade_factory_snapshot_set_property_cb,
+};
+
+
+void deploy_task_handle_exit_status(struct DeployTask *task, int exit_status)
+{
+    GError *error = NULL;
+    if (!g_spawn_check_exit_status(exit_status, &error)) {
+        if (task->checked) {
+            SFMF_FAIL("Failed to run command: %s\n", error->message);
+        } else {
+            SFMF_WARN("Failure (ignored): %s\n", error->message);
+        }
+        g_error_free(error);
+    }
+}
+
+struct DeployTask *deploy_task_queue_get_current_task(struct DeployTaskQueue *queue)
+{
+    if (queue->current >= 0) {
+        return &(queue->tasks[queue->current]);
+    }
+
+    return NULL;
+}
+
+void deploy_task_queue_run_sync(struct DeployTaskQueue *queue)
+{
+    struct DeployTask *task = deploy_task_queue_get_current_task(queue);
+
+    if (!task) {
+        SFMF_FAIL("Current task is invalid\n");
+    }
```

Comment:
`SFMF_FAIL` should always exit, maybe I should fix up the places where code assumes it returns (that'll be code that never gets called).